### PR TITLE
Adding file renyi_entropy.py for calculating renyi entropies (Resolves #1088)

### LINKED ISSUE #1088: Implement the Rényi entropy of order $\alpha$ of a quantum state
The Rényi entropy of order $\alpha$ of a PSD matrix $\rho$ is defined as
$$H_{\alpha}(\rho)=\frac{1}{1-\alpha}\log\left(\frac{\mathrm{Tr}\left[\rho^{\alpha}\right]}{\mathrm{Tr}[\rho]}\right)$$
for all $\alpha\in(0, 1)\cup(1,+\infty)$. It can be extended in $\alpha=1$, resulting in the Von Neumann entropy, and in $\alpha=+\infty$, resulting in the min-entropy, as seen in [1306.3142](https://arxiv.org/abs/1306.3142).

These entropies are more and more studied in finite-size quantum cryptography. As such, it could be interesting to include them in toqito.

### PR DESCRIPTION
**Pull request by GauravR-Malik**
## Description

This pull request includes a .py file to calculate the renyi entropy of a given state provided as a density matrix. It is an attempt that Resolves #1088  

## Changes

  -  [x] Change 1 Adds a new file in the folder state_ops within the toqito folder in the repository called renyi_entropy.py
  -  [x] Change 2 Adds a new file in the tests folder within state_ops called as test_renyi_entropy.py with test cases
  -  [x] Change 3 Modifies the existing __init__.py to initialize the renyi_entropy

## Checklist
Before marking your PR ready for review, make sure you checked the following locally. If this is your first PR, you might be notified of some workflow failures after a maintainer has approved the workflow jobs to be run on your PR. 

Additional information is available in the [documentation](https://toqito.readthedocs.io/en/latest/contributing.html#testing).

  -  [x] Use `ruff` for errors related to code style and formatting.
  -  [x] Verify all previous and newly added unit tests pass in `pytest`.
  -  [x] Check the documentation build does not lead to any failures. `Sphinx` build can be checked locally for any failures related to your PR
  -  [x] Use `linkcheck` to check for broken links in the documentation
  -  [x] Use `doctest` to verify the examples in the function docstrings work as expected.

### CODE REVIEW

**Comment by GauravR-Malik** [comment]
Hi @vprusso. I have just created a pull request for resolving issue #1088. Could you please assign me this?
I am still a begineer in contributing to open source via GitHub... if there are any mistakes from my side in the workflow, do let me know. I'd look forward to resolve them.

**Comment by vprusso** [comment]
@GauravR-Malik all I see in your commit is two lines:
https://github.com/vprusso/toqito/pull/1102/files

You might want to read the docs on how to use git and GitHub. Maybe something like [this](https://www.youtube.com/watch?v=Oaj3RBIoGFc) would be helpful for you.

**Comment by GauravR-Malik** [comment]
Oh yes. The commit is just an added file with a few lines in comments... I do need to add the lines for making the function work. But wouldn't you have to assign the issue to me first, before I start adding in the content? I was waiting for that...

**Comment by vprusso** [comment]
> Oh yes. The commit is just an added file with a few lines in comments... I do need to add the lines for making the function work. But wouldn't you have to assign the issue to me first, before I start adding in the content? I was waiting for that...

You don't need to be assigned to push to your branch.

**Comment by GauravR-Malik** [comment]
Yes. I get that. I can add commits to my own branch. But shouldn't I be assigned the issue before I start working on it? It seemed like that was the process... Am I missing something? Do let me know... I am sorry to be bothering you again, but it seems fine at my end.

**Comment by vprusso** [comment]
> Yes. I get that. I can add commits to my own branch. But shouldn't I be assigned the issue before I start working on it? It seemed like that was the process... Am I missing something? Do let me know... I am sorry to be bothering you again, but it seems fine at my end.

@GauravR-Malik I can't make any meaningful comments on your MR unless you submit what you are working on. You don't need to be assigned to make an MR. Additionally, as mentioned on other issues that you commented on, due to the volume of individuals interested in contributing, it makes more sense to have an earnest attempt at an issue first as I've assigned numerous issues to others only to have them not complete the task, etc.

**Comment by GauravR-Malik** [comment]
Alright. Sounds Fair. I will add the commits to complete this function first.

**Comment by GauravR-Malik** [comment]
Hi @vprusso I have added the lines for computing the values of the renyi entropy

File: toqito/state_ops/renyi_entropy.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+# Includes code to calculate the renyi entropy of a given state, as a wavefunction or a density matrix
+# Takes into account the special cases of \alpha = 0,1 or \infty
```
Comment (vprusso) [inline toqito/state_ops/renyi_entropy.py:2]: Please refer to other files in `toqito` to see how the header comments are formatted.

File: toqito/state_ops/renyi_entropy.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+import numpy as np
+
+# Defining the function for Renyi Entropy
```
Comment (vprusso) [inline toqito/state_ops/renyi_entropy.py:6]: This comment is not useful as this is implicit.

File: toqito/state_ops/renyi_entropy.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+def renyi(rho,alpha):
```
Comment (vprusso) [inline toqito/state_ops/renyi_entropy.py:7]: The name of the function should be the name of the file. You also need to type-hint the variables and the return type of the function.

File: toqito/state_ops/renyi_entropy.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    """
+    Calculate the Rényi entropy for a quantum density matrix rho with parameter alpha.
```
Comment (vprusso) [inline toqito/state_ops/renyi_entropy.py:9]: You need to provide a reference. Please refer to other functions in toqito for how to do this.

File: toqito/state_ops/renyi_entropy.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    
+    Parameters:
+    -----------
```
Comment (vprusso) [inline toqito/state_ops/renyi_entropy.py:12]: Refer to how docstrings are formatted elsewhere in the project. The type information should be coming from the type-hints, and not defined in the docstring.

File: toqito/state_ops/renyi_entropy.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        
+    Returns:
+    --------
+    float
+        The Rényi entropy value
+    """
+  
+    rho = np.array(rho)
+    eigenvalues = np.linalg.eigvalsh(rho)
+    eigenvalues = eigenvalues[eigenvalues > 1e-10]
```
Comment (vprusso) [inline toqito/state_ops/renyi_entropy.py:26]: The tolerance param should be an optional argument of the function and not defined explicitly here.

File: toqito/state_ops/renyi_entropy.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+    # First resolving the special values of alpha
+    if(alpha == 0):
```
Comment (vprusso) [inline toqito/state_ops/renyi_entropy.py:29]: Putting parens around an expression like this is not the way that things are typically formatted in Python. You also should have a space after the `if`.

**Review by vprusso — COMMENTED** [review]
@GauravR-Malik you also need to add unit tests for this function along with including it in the `__init__.py` file.

**Comment by GauravR-Malik** [comment]
Sure @vprusso. Shall add the unit tests and make the other improvements. Thanks again for your time.

**Comment by vprusso** [comment]
> Sure @vprusso. Shall add the unit tests and make the other improvements. Thanks again for your time.

Yes, please.

**Comment by vprusso** [comment]
> Hi @vprusso. Just an update. Was really involved in a few other things during the last two days and couldn't work on adding the remaining tasks. Would finalise the pull request in a day or two.

No problem, thank you for the update.

**Comment by GauravR-Malik** [comment]
Hi @vprusso I have made the required commits in this pull request to add the test cases and have also modified the __init__.py file. The ruff and doctest run in my local machine also do not report any failures. Could you review the pull request again? I feel it is complete. If there is something that is left out, I'd be eager to get it right. Thanks.

File: docs/refs.bib [hunk 8]
```diff
@@ -765,6 +765,17 @@ @misc{Molina_2012_Optimal
   primaryClass={quant-ph}
 }
 
+@article{muller_lennert_renyi_2013,
```
Comment (vprusso) [inline docs/refs.bib:768]: Please refer to the other references in this file for how to format the bibtex key.

File: docs/refs.bib [hunk 9]
```diff
@@ -870,6 +881,14 @@ @article{Pusey_2012_On
 }
   
 % Last name begins with Q
+@misc{quantiki_entropy,
```
Comment (vprusso) [inline docs/refs.bib:884]: Similar comment to the one above here.

File: toqito/state_ops/renyi_entropy.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        renyi = (-1) * np.sum(np.log2(eigenvalues) * eigenvalues)
+
+    elif np.isinf(alpha):
+        renyi = (-1) * np.log2(np.max(eigenvalues))
+
+    else:
+        pow_eigvals = np.power(eigenvalues, [alpha])
+        renyi = np.log2(np.sum(pow_eigvals)) / (1 - alpha)
+
+    return float(abs(renyi))
```
Comment (vprusso) [inline toqito/state_ops/renyi_entropy.py:74]: Perhaps this is a bit cleaner? 

```suggestion

    if alpha == 0:
        return np.log2(len(eigenvalues))

    if alpha == 1:
        return -np.sum(eigenvalues * np.log2(eigenvalues))

    if np.isinf(alpha):
        return -np.log2(np.max(eigenvalues))

    # General alpha case
    sum_power = np.sum(eigenvalues ** alpha)
    renyi = np.log2(sum_power) / (1 - alpha)

    return float(renyi)
```

File: toqito/state_ops/tests/test_renyi_entropy.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
+"""Test Renyi Entropy."""
+
+import numpy as np
+import pytest
+
+from toqito.state_ops import renyi_entropy
+
+# Example density matrices for testing
```
Comment (vprusso) [inline toqito/state_ops/tests/test_renyi_entropy.py:8]: Can you please use `pytest.parameterize` to cycle over these different cases and consolidate the tests? You can look at the other tests that use `parameterize` in toqito for reference.

File: toqito/state_ops/tests/test_renyi_entropy.py [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+def test_alpha_inf():
+    """Testing the Renyi Entropy of order infty for the pure state and mixed."""
+    assert renyi_entropy(rho_pure, np.inf) == pytest.approx(0.0)
+    assert renyi_entropy(rho_mixed, np.inf) == pytest.approx(1.0)
+
+
+def test_invalid_density_matrix():
+    """Testing the Renyi Entropy of incorrect density matrix as trial input."""
+    with pytest.raises(ValueError, match="The density matrix must have trace equal to 1"):
+        renyi_entropy(invalid_rho, 1)  # Should raise ValueError
```
Comment (vprusso) [inline toqito/state_ops/tests/test_renyi_entropy.py:35]: I'm not implying that you did (nor do I personally care), but many of these tests seem like they were generated by an LLM. You should ensure that not only is the coverage of the tests complete, but that the format of the tests as written abide by and align with the other tests in toqito.

File: toqito/state_ops/renyi_entropy.py [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+          .. math:: S_1(\rho) = -\sum_i \lambda_i \log_2 \lambda_i.
+        - For :math:`\alpha \to \infty` (Min-entropy):
+          .. math:: S_{\infty}(\rho) = -\log_2 \max_i \lambda_i.
+
+
+    Examples
+    ========
+    Compute the Rényi entropy of a pure state:
+
+    >>> import numpy as np
```
Comment (vprusso) [inline toqito/state_ops/renyi_entropy.py:30]: Won't these examples fail without an import for renyi_entropy?

File: toqito/state_ops/renyi_entropy.py [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+
+    :param rho: The quantum density matrix (must be Hermitian, positive semi-definite, and have trace 1).
+    :param alpha: The order of Rényi entropy.
+    :param tolerance: The numerical tolerance for eigenvalues (default is :math:`10^{-10}`).
+    :raises ValueError: If the density matrix does not have trace equal to 1.
+    :return: The Rényi entropy value.
+
+    """
+    if not np.allclose(np.trace(rho), 1.0, atol=1e-10):
+        raise ValueError("The density matrix must have trace equal to 1")
```
Comment (vprusso) [inline toqito/state_ops/renyi_entropy.py:55]: You could just use `toqito.matrix_props.is_density` to check this instead.

File: toqito/state_ops/renyi_entropy.py [hunk 15]
```diff
@@ -0,0 +1,74 @@
+"""Calculates the renyi entropy for a given density matrix. Valid for special cases of alpha = 0,1 or infty."""
```
Comment (vprusso) [inline toqito/state_ops/renyi_entropy.py:1]:
```suggestion
"""Calculates the renyi entropy for a given density matrix."""
```

File: toqito/state_ops/tests/test_renyi_entropy.py [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+
+import numpy as np
+import pytest
+
+from toqito.state_ops import renyi_entropy
+
+# Example density matrices for testing
+rho_pure = np.array([[1, 0], [0, 0]])  # Taking a pure state density matrix
+rho_mixed = np.array([[0.5, 0], [0, 0.5]])  # Taking a maximum mixed state
+invalid_rho = np.array([[0.5, 0], [0, 0.4]])  # Taking error case, trace not equals 1
```
Comment (vprusso) [inline toqito/state_ops/tests/test_renyi_entropy.py:11]: Are there known cases in the literature that would be useful or interesting to include as test cases? My fear is that only basic mixed or pure states will not be sufficient for capturing potential edge cases.

File: toqito/state_ops/renyi_entropy.py [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+    where :math:`\lambda_i` are the eigenvalues of :math:`\rho`.
+
+    Special cases :cite:`muller_lennert_renyi_2013`
+        - For :math:`\alpha = 0` (Hartley entropy):
+          .. math:: S_0(\rho) = \log_2 d,
+          where :math:`d` is the rank of :math:`\rho`.
+        - For :math:`\alpha = 1` (Shannon entropy):
+          .. math:: S_1(\rho) = -\sum_i \lambda_i \log_2 \lambda_i.
+        - For :math:`\alpha \to \infty` (Min-entropy):
+          .. math:: S_{\infty}(\rho) = -\log_2 \max_i \lambda_i.
```
Comment (purva-thakre) [inline toqito/state_ops/renyi_entropy.py:23]: This is not formatted correctly. 

image

https://toqito--1102.org.readthedocs.build/en/1102/autoapi/state_ops/renyi_entropy/index.html

Please check the contributing guide on how you can verify documentation examples etc. locally.

**Comment by GauravR-Malik** [comment]
Thanks for your reviews @vprusso and @purva-thakre. I do agree about the test cases being generic... I'd lookup some edge cases and add them as well. Meanwhile I am resolving the other issues also. Thanks for your time with this review.

**Comment by purva-thakre** [comment]
@GauravR-Malik Is this PR ready for a review?

**Comment by GauravR-Malik** [comment]
Hi @purva-thakre. Not yet actually... I still have to make some changes and add some relevant examples... I was involved elsewhere, and couldn't complete. 
I will add the changes soon and then ask for another review.

**Comment by vprusso** [comment]
Just want to flag that some of the Renyi entropy stuff may be better handled by the recent work done in the `qics` conic quantum information solver package:
https://qics.dev/examples/qrep/renyi.html

(FYI @purva-thakre / @GauravR-Malik )

**Comment by tnemoz** [comment]
> Just want to flag that some of the Renyi entropy stuff may be better handled by the recent work done in the `qics` conic quantum information solver package: https://qics.dev/examples/qrep/renyi.html
> 
> (FYI @purva-thakre / @GauravR-Malik )

@vprusso Wouldn't that be more relevant for #1089?

**Comment by vprusso** [comment]
> > > Just want to flag that some of the Renyi entropy stuff may be better handled by the recent work done in the `qics` conic quantum information solver package: https://qics.dev/examples/qrep/renyi.html
> > > (FYI @purva-thakre / @GauravR-Malik )
> > 
> > 
> > @vprusso Wouldn't that be more relevant for #1089?
> 
> Yes, I think it would be _more_ relevant to flag under #1089, but possibly could also be considered here as well? In any case, I definitely agree that putting this in the description of #1089 is sensible, so let me go ahead and do that. Thank you for flagging that, @tnemoz !

**Comment by purva-thakre** [comment]
@GauravR-Malik Are you still interested in wrapping up this PR?

**Comment by GauravR-Malik** [comment]
> @GauravR-Malik Are you still interested in wrapping up this PR?

Yes indeed. I had to take a break from the PR due to some work which had turned up last month... but I shall be on it in the coming week, and wrap it up soon.
